.pylintrc

```diff
@@ -69,7 +69,9 @@ disable=no-self-use,        # disabled as it is too verbose
     too-many-public-methods, too-few-public-methods, too-many-ancestors,
     unnecessary-pass,   # allow for methods with just "pass", for clarity
     no-else-return,     # relax "elif" after a clause with a return
-    docstring-first-line-empty # relax docstring style
+    docstring-first-line-empty, # relax docstring style
+    cyclic-import # This is unreliable and often returns false positives
```

Comment:
I don't think we should just disable this. I'd err on the side of being conservative with these things as it enforces some discipline on our namespace usage.

In your case in particular there's really no need, just do your imports from `qiskit.circuit`, rather than the root, and the lint warning goes away.

Reply:
Sure, you can workaround the failures, but we shouldn't have to. There is no cycle here despite what pylint thinks. If there was an actual import cycle python would complain. Switching to import from `qiskit.circuit` instead of `qiskit` is good practice in general yes, but first the error message doesn't give any hints as to that (see the commit message on: https://github.com/Qiskit/qiskit-terra/pull/2137/commits/7e614ec6826a95eabbe69a4c40474031bd6c58a0 ). But secondly and more importantly the way we setup the root namespace isn't the best (see #1170 ) and that's the only reason pylint keeps throwing these errors. But instead of telling us what the actual problem is it throws cryptic errors from this check which is why I disabled it.

I keep having to reference #1179 as all the constant issues with lint checks are just symptomatic of how poor a tool pylint is for how we're using it. Even if you ignore how fragile it is from a dependency standpoint and that things outside of its dependency list constantly break us (or how I can't run it locally on 3.7 because of that). In general it does more harm then it helps. Just because there is a rule in pylint does not mean it provides us actual value. If we're actually concerned about our development throughput switching to a tool that works reliably (both in CI and locally), runs more quickly, and has a sane rule set will do far more for that than practically anything else.

Reply:
I'm happy to reevaluate our linter after the release. But this PR should not be tied to that, especially because there is no need for this to be disabled here. You already agree that importing from `qiskit.circuit` is better.

FWIW, I don't think the message is cryptic at all. It says:
```
Cyclic import (qiskit -> qiskit.execute -> qiskit.compiler -> qiskit.compiler.disassembler) (cyclic-import)
```
There is a clear loop here; it just happens that Python can resolve it at runtime. In contrast, when Python fails on a cyclic import, it just says "cannot import", which is far more cryptic.

Static checks are heuristic and may be imperfect. However, disabling them in favor of just relying on a runtime error is not a great idea. For one, your tests may not have 100% coverage and you may not hit the runtime error in your tests.

In this particular case, it is actually highlighting our abuse of the root namespace, which I think you agree with. We have shoved many things arbitrarily in there, and I'd rather clean that up than circumvent it by disabling static checks.